docs/testing.md

```diff
@@ -0,0 +1,138 @@
+# Testing BMI
+
+Every Developer will need to test his/her feature or make sure that he/she
+did not break anything. So the test cases are situated in
+
+```
+/tests/
+```
+
+## Directory Structure
+
+This folder follows the directory structure of the project with additional
+folders like for example operation.py's test cases are situated in
+```
+/tests/einstein/operations/unit/...
+/tests/einstein/operations/integration/...
+```
+
+Every file must have a folder for itself which has additional folders called
+'unit' and 'integration'
+
+All the unit tests should be in unit folder and all integration tests should
+be in integration folder for that file
+
+All of these folders are python packages so they should have a \_\_init\_\_.py
+
+## Writing Tests
+
+Create a file in the correct location of the directory structure and make sure
+the name conforms to the regex 'test_\*.py'
+
+We are using python's unittest as our testing framework for now.
+
+So the file can have any number of test cases as long as it makes sense they
+belong together.
+
+Each test case is represented by a class and should extend unittest.TestCase
+
+This is the typical structure of a test case
+
+```
+from ims.common.log import trace
+
+class TestSample(unittest.TestCase):
+
+  @trace
+  def setUp(self):
+       pass
+
+  def runTest(self):
+       pass
+
+  def tearDown(self):
+       pass
+ ```
+ 
+* Each class's name should start with Test
+* Each setUp method should use the trace decorator
+* setUp method is supposed to do stuff which is required before the actual test
+is run like adding db entries, etc.
```

Comment:
I didnt add it because it is not a definitive list, it depends on the test.
Like for example the provision and deprovision test cases are actually same, but in provision the deprovision call goes into teardown and provision goes into setup in deprovision test case.